CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java

```diff
@@ -0,0 +1,517 @@
+											
+											rules = chooseRules(rules, provider, declaringClassName);
+											break outerloop;
+										}
+									}
+								}
+							}
+						}
+					}
+				}	
+			}
+		}
+	 			
+		return rules;
+	}
+	
+	
+	
+	// Methods used from the `doAnalysis()` method
+	//-----------------------------------------------------------------------------------------------------------------
+	
+	/**
+	 * This method returns the type of Provider detected, since
+	 * it can be either `java.security.Provider` or `java.lang.String`
+	 * 
+	 * @param providerValue
+	 */
+	private String getProviderType(Value providerValue) {
+		String providerType = providerValue.getType().toString();
+		return providerType;
+	}
+	
+	
+	/**
+	 * This method return the provider used when Provider detected is of type `java.security.Provider`
+	 * 
+	 * @param statement
+	 *            
+	 * @param sootMethod
+	 *           
+	 * @param providerValue
+	 *            
+	 * @param icfg
+	 *            
+	 */
+	private String getProviderWhenTypeProvider(JAssignStmt statement, SootMethod sootMethod, Value providerValue, ObservableICFG<Unit, SootMethod> observableDynamicICFG) {
+		String provider = null;
+		
+		//Create a Boomerang solver.
+		Boomerang solver = new Boomerang(new DefaultBoomerangOptions(){
+			public boolean onTheFlyCallGraph() {
+				//Must be turned of if no SeedFactory is specified.
+				return false;
+			};
+		}) {
+			@Override
+			public ObservableICFG<Unit, SootMethod> icfg() {
+				return observableDynamicICFG;
+			}
+
+			@Override
+			public SeedFactory<NoWeight> getSeedFactory() {
+				return null;
+			}
+		};
+		Map<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> map = Maps.newHashMap();
+		for(Unit pred : observableDynamicICFG.getPredsOf(statement)) {
+			//Create a backward query
+			BackwardQuery query = new BackwardQuery(new Statement((Stmt) pred,sootMethod), new Val(providerValue, sootMethod));
+			//Submit query to the solver.
+			
+			BackwardBoomerangResults<NoWeight> backwardQueryResults = solver.solve(query);
+			map.putAll(backwardQueryResults.getAllocationSites());
+		}
+		
+		// The Provider can be correctly detected from this static analysis, if there is only one allocation site
+		// where the Provider variable was initialized. Otherwise, it throws an error because it is not possible
+		// to detect for sure the provider, if is given as parameter to the getInstance() method through the use of
+		// IF-ELSE, SWITCH statements or TERNARY operators
+		if(map.size() == 1) {
+			for(Entry<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> entry : map.entrySet()) {
+				ForwardQuery forwardQuery = entry.getKey();
+				
+				Val forwardQueryVal = forwardQuery.var();
+				Value value = forwardQueryVal.value();
+				Type valueType = value.getType();
+				String valueTypeString = valueType.toString();
+				
+				if(valueTypeString.contains("BouncyCastleProvider")) {
```

Comment:
Yes, the detection only support BC provider. Initially, I wanted to design my provider detection analysis differently. For example, in the statement `Provider p1 = new BouncyCastleProvider();`,  since the `p1` variable holds the value `BC`, I wanted to get that value at the end of my provider analysis.

But I was not able to do this when using Boomerang, and therefore used the current approach. I am getting the type of the `p1` and checking what it is. Currently, I am supporting only BC provider. Furthermore, if we would want to support another provider, I just need to update this section according to the new provider. Feel free to tell me if a better approach comes to mind :slightly_smiling_face: